SPT: immediately render modal with empty thumbnails

retrofox opened the pull request:
#### Changes proposed in this Pull Request

This PR changes slightly the approach in how the thumbnails are dynamically previewed.

_Although it doesn't depend strictly on https://github.com/WordPress/gutenberg/pull/17242, it will improve reduce the previewing time even for all thumbnails and as well as the large preview (100ms)_

#### Overview

The current behavior waits to get the templates parsed before to start to render and mount the components. It produces a delay when the template modal appears. In order to improve the performance and the user perception, I purpose the following more important changes in this PR:

* `Disengage` the parsing process from the react component, starting to process as fast as possible without waiting for the component lifecycle.
* `Do not update the state` with the parsed templates blocks. Update the state[ seems to take considerable time.](https://github.com/Automattic/wp-calypso/pull/35713#issuecomment-536676899)
* Add a `templates-parser.js` which acts as a data store and global event dispatcher. It contains selectors in order to get the desired data.

* `Dispatch onTemplateParse` global event to communicate with the component. The event dispatcher sends all data about templates in the event object argument:

```es6
window.addEventListener( 'onTemplateParse', event => console.log ( event ) );
```

![image](https://user-images.githubusercontent.com/77539/66085323-41bf3700-e547-11e9-8a5d-dbafb3c050b7.png)

The event detail object contains the following fields:

* `blocks`: list of the parsed blocks
* `count`: the number of blocks
* `isEmpty`: `true` is the template doesn't have content.
* `isParsing`: defines if the template is being parsed or not.
* `title`: template title.
* `slug`: template slug.

#### Testing instructions

Apply the patch and confirm that you see the spinners in the thumbnails when the templates are being parsed.

#### Static vs Dynamic


#### Tests


##### Parsing One By One
```
time - blank: 0.005126953125ms
time - about: 46.718017578125ms
time - contact: 19.9072265625ms
time - portfolio: 28.283935546875ms
time - services: 51.375ms
time - team: 37.27197265625ms
time - blog: 29.65673828125ms
time - menu: 51.2978515625ms
time - total: 6304.18798828125ms
```

obenland commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/template-selector-item.js:
```diff
@@ -26,6 +26,10 @@ const TemplateSelectorItem = props => {
 		blocks = [],
 	} = props;
 
+	if ( ! blocks.length && 'blank' !== value ) {
```
Rather then checking for blocks here, how do you feel about testing `this.state.blocks` in `PageTemplateModal.render()`?

```js
if ( ! this.state.isOpen || isEmpty( this.state.blocks ) ) {
		return null;
}
```

getdave replied:
I still think the component should function when the props aren't provided as expected. It makes them more resilient when moved to an alternative use case. 

This doesn't preclude your change happening as well.

retrofox committed:
spt: wait to parse template blocks

retrofox committed:
spt: rendering thumbnails one by one

getdave commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/template-selector-item.js:
```diff
@@ -26,18 +26,27 @@ const TemplateSelectorItem = props => {
-	);
+	const renderInnerButton = () => {
+		if ( ! blocks.length ) {
+			return 'blank' !== value ? <Spinner /> : null;
```
Just wondering whether this component should be aware of the concept of `blank`?

retrofox replied:
yes, it's a good point. Let's find a way to remove that checking.

getdave commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -47,20 +47,23 @@ class PageTemplateModal extends Component {
+						? parseBlocks( replacePlaceholders( content, siteInformation ) )
+						: [];
+					this.setState( { blocks: { ...this.state.blocks, [ slug ]: parsedBlocks } } );
+				} ).bind( null, templates[ i ] ),
```
This is capturing the template at index?

retrofox replied:
Not sure if I'm understanding the question rightly, but we are doing with that is passing the template instance at the scope of the timeout callback.

getdave commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/styles/starter-page-templates-editor.scss:
```diff
@@ -146,10 +146,12 @@ body.admin-bar:not( .is-fullscreen-mode ) .page-template-modal-screen-overlay {
+			position: absolute;
+			top: 50%;
+			left: 50%;
+			margin: 0 0 0 -9px;
```
What's the `-9px` magic number?

retrofox replied:
:magic: :-D the spinner width is 18px, so it moves the spinner at left with the half this value.

getdave replied:
> :magic: :-D

😄 Sorry. Not sure if everyone is aware of this as a "thing" but Harry Roberts has this down as one of his things to look out for in CSS codebase reviews

https://csswizardry.com/2012/11/code-smells-in-css/#magic-numbers

> ...the spinner width is 18px

A small point but I have found it is helping to have comments explaining why these numbers exist or use well named Sass variables to make the code self-documenting. What do you think? Is it overkill?

```
margin: 0 0 0 -#{$spinner-width / 2};
```

getdave reviewed (commented):
Async rendering these is a massive win. Great workl

retrofox committed:
spt: remove `blanck` template checking

getdave commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:
```diff
@@ -61,7 +61,7 @@ class PageTemplateModal extends Component {
 					this.setState( { blocks: { ...this.state.blocks, [ slug ]: parsedBlocks } } );
 				} ).bind( null, templates[ i ] ),
-				500 * i
+				50 * i
```
I'd be for capturing this value in a well-named variable to make the code self-documenting.

retrofox replied:
Adding. No guarantee with _well-named_ 😅

retrofox committed:
spt: better support for themes

retrofox committed:
spt: add a constant to define parsing delay

retrofox committed:
set scaling delay to 0

retrofox committed:
spt: render empty buttons when no blocks
squash: add isTemplateLoadingBySlug to state
enable dynamic preview. propagate template parsing state
show spinner when still parsing
spt: adjust thumbs spinner position

obenland commented:
Page load times are still extremely long. Locally it takes 9s to fully load the page, including the thumbnail previews. With static thumbnails it's 3s (which is still long).

I'm afraid we'll need to do better before this is mergeable. :/

obenland commented:
@youknowriad Do you happen to have any ideas how we can cut down on the time it takes to setState here?

![Screen Shot 2019-09-30 at 10 53 07 AM](https://user-images.githubusercontent.com/1398304/65903068-965c8980-e370-11e9-8e68-39f97f090b28.png)

retrofox commented:
> Locally it takes 9s to fully load the page,

it doesn't take that time to me. Let me paste a gif with a clock at the bottom right corner.

![dynamic-preview-02](https://user-images.githubusercontent.com/77539/65903403-bef57b00-e392-11e9-8eba-ba7fa4c34f6d.gif)

retrofox commented:
I've measured both process. The [reduce loop](https://github.com/Automattic/wp-calypso/pull/35713/files#diff-a664d83ace51cb87304bf656c9e9f303R79-R91) wich contemplates parsing all block of all templates: `time:reduce: 351.2021484375ms`,
and setting the state: `time:set-state: 1388.221923828125ms`

retrofox commented:
Another thing I forgot to mention that is I'm testing this PR together with [this one](together ), reducing the delay before to start to compute the scale factor from 100ms to 0, through of the `__exoperimentalScalingDelay` property.

retrofox committed:
add tempates-parser.js

retrofox committed:
get first template slug selector

retrofox committed:
remove constructor

youknowriad commented:
> @youknowriad Do you happen to have any ideas how we can cut down on the time it takes to setState here?

I assume that's the setState that sets the block list and triggers the rendering. I fear that there's not much we can do there. things to consider maybe:

 - Cut the number of blocks to render (or maybe render some and then update the list with the full list async)
 - In Core an idea to explore is a "preview mode" for blocks where maybe somethings can be removed entirely, components only useful for edit like block appenders, movers, dropzones... (it needs to be measured though to avoid unnecessary complexity)

retrofox commented:
_Preview mode_ is something that worths to dive imo. [We talked about this a few weeks ago](https://wordpress.slack.com/archives/C02QB2JS7/p1566302350383400).
What I've done, just for this approach, is avoiding to write the state with the parsed blocks. Going to update the PR description.

retrofox committed:
adjust title in large preview

retrofox committed:
dispatch onTemplatesParse global event.

getdave commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/templates-parser.js:
```diff
@@ -0,0 +1,81 @@
+	title,
+} ) );
+
+// Dispatch a global event once all templates have been parsed.
```
Is this comment accurate? When I first saw it I didn't understand what was going on here. Maybe:

> Dispatch a global event to indicate templates are being parsed

getdave commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/templates-parser.js:
```diff
@@ -0,0 +1,81 @@
+			},
+		} )
+	);
+}, 50 );
```
So we're still effectively parsing in serial? So the main benefit is that we render something to the page immediately and then once the templates are done parsing we "send" that data to the component to re-render with the actual previews.

getdave commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/utils/templates-parser.js:
```diff
@@ -0,0 +1,81 @@
+	);
+}, 50 );
+
+export const getAllTemplatesBlocks = () => allTemplatesBlockBySlug;
```
It's great that you wrote all of these otherwise they'd be littered all of the place. Nice work 👍

getdave reviewed (commented):
Great work continuing with this. It is so much better!

~I couldn't see any spinners but I assume this is due to not applying https://github.com/WordPress/gutenberg/pull/17242~

I activated [Balasana](https://wordpress.com/theme/balasana) and now I see spinners. I wonder why? 

Some other code comments.

If this has been measured objectively with timers to be an improvement then it's clearly a win, because getting _anything_ onto the page in front of the user is better than waiting for the entire process to complete before rendering.

Thanks for persevering here. Let me know if/how I can help.

retrofox commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/block-template-preview.js:
```diff
@@ -14,16 +14,18 @@ import { BlockPreview } from '@wordpress/block-editor';
+						<BlockPreview
+							blocks={ blocks }
+							viewportWidth={ viewportWidth }
+							__experimentalScalingDelay={ 0 }
```
Yes, but it isn't a blocker. Agree we need to keep an eye open there but it won't break anything if WordPress/gutenberg#17242 ships.

obenland commented:
The current version performs pretty well in Chrome for me, still a rather long delay in Safari.
Have you noticed a long phase of white screen before the page displays (in either browser)?

retrofox commented:
> Have you noticed a long phase of white screen before the page displays (in either browser)?

Does it happen even when you add a new post? It seems to be the default behavior when we should the core-editor in the wp-admin.

retrofox committed:
spt: render templates, one by one.

retrofox committed:
spt: preview styles improvements

lancewillett commented:
Cleaning up older pull requests as part of a GitHub scrub.

Is this change still valid? What's needed to move this along? It looks like an important performance boost. CC @retrofox 

If not valid, please close with a comment. Thanks!

obenland commented:
There's a new version of SPT out now, let's close this.
